### PR TITLE
testing: Add some more debug scaffolding around the hotrestart_test

### PR DESCRIPTION
**Pull request by jmarantz**
*Description*:
hotrestart_test is difficult to debug when it goes wrong, because there are many different sub-tests within it, and it's hard to track where a failure occurs.  This borrows some sh-test debug scaffolding from PageSpeed where it's proved useful.

*Risk Level*: Low

*Testing*:
//test/integration:hotrestart_test
//test/integration:hotrestart_test --runs_per_test=10 --jobs=1
//test/integration:hotrestart_test --config=tsan --runs_per_test=10 --jobs=1

*Release Notes*: N/A

### CODE REVIEW

File: test/integration/hotrestart_test.sh [hunk 1]
```diff
@@ -2,6 +2,77 @@
...
+# infrastructure, where are part of a bash system-test helper library.
+# I'm putting them here for this one test to help me debug it, with
+# the thinking that if there's traction for this infrastructure I'll
+# factor it out into a helpers class for Envoy.  Original source link:
+#
+# https://github.com/apache/incubator-pagespeed-mod/blob/c7cc4f22c79ada8077be2a16afc376dc8f8bd2da/pagespeed/automatic/system_test_helpers.sh#L383
+
+CURRENT_TEST="NONE"
+function start_test() {
+  #update_elapsed_time
```
Comment (ggreenway) [inline test/integration/hotrestart_test.sh]: Delete commented-out stuff?

File: test/integration/hotrestart_test.sh [hunk 2]
```diff
@@ -48,23 +120,22 @@ fi
...
+  start_test Starting epoch 0
   ADMIN_ADDRESS_PATH_0="${TEST_TMPDIR}"/admin.0."${TEST_INDEX}".address
-  "${ENVOY_BIN}" -c "${HOT_RESTART_JSON}" \
+  run_in_background_saving_pid "${ENVOY_BIN}" -c "${HOT_RESTART_JSON}" \
       --restart-epoch 0 --base-id 1 --service-cluster cluster --service-node node \
-      --admin-address-path "${ADMIN_ADDRESS_PATH_0}" &
+      --admin-address-path "${ADMIN_ADDRESS_PATH_0}"
 
-  FIRST_SERVER_PID=$!
-  sleep 3
```
Comment (ggreenway) [inline test/integration/hotrestart_test.sh]: Without some time here (not that I love arbitrary sleep in test...), couldn't we send signals to the process before it has installed signal handlers?

File: test/integration/hotrestart_test.sh (hunk 1 above)
Comment (jmarantz) [inline test/integration/hotrestart_test.sh]: done

File: test/integration/hotrestart_test.sh (hunk 2 above)
Comment (jmarantz) [reply test/integration/hotrestart_test.sh]: good catch; done.

**Comment by ggreenway** [comment]
Need to resolve merge conflicts

File: test/integration/hotrestart_test.sh [hunk 3]
```diff
@@ -2,6 +2,75 @@
...
+  local i
+  local stack_size=${#FUNCNAME[@]}
+  for (( i=1; i<$stack_size ; i++ )); do
+    local func="${FUNCNAME[$i]}"
+    [ -z "$func" ] && func=MAIN
+    local line_number="${BASH_LINENO[(( i - 1 ))]}"
+    local src="${BASH_SOURCE[$i]}"
+    [ -z "$src" ] && src=non_file_source
+    local canonical_dir=$(cd $(dirname "$src") && pwd)
+    local short_dir=${canonical_dir#*/net/instaweb/}
```
Comment (htuch) [inline test/integration/hotrestart_test.sh]: LGTM if you can cleanup this stuff.

**Comment by ggreenway** [comment]
More merge conflicts

**Comment by dnoe** [comment]
@htuch for final review pass

**Pull request merged**